File: WhatsApp-agents/whatsmeow/internal/contracts/env.go
package contracts

import (
	"fmt"
	"net/url"
	"os"
	"strconv"
	"strings"
)

// AppEnv - validated configuration used across layers after startup parse.
type AppEnv struct {
	StoreDir          string
	Port              int
	CouponMediaURL    string
	MenuMode          string
	WebhookSecret     string
	GoEnv             string
	HumanizeMinMS     int
	HumanizeMaxMS     int
}

// EnvParseResult - either validated config or a clear invalid-keys error.
type EnvParseResult struct {
	OK    bool
	Env   AppEnv
	Error string
}

// ParseEnv - fail-fast env validation (defaults, production delay floor).
func ParseEnv(raw map[string]string) EnvParseResult {
	storeDir := pick(raw, "WHATSMEOW_STORE_DIR", "./whatsmeow_store")
	if strings.TrimSpace(storeDir) == "" {
		return fail("WHATSMEOW_STORE_DIR: must be non-empty")
	}

	port, err := parseInt(pick(raw, "PORT", "3004"))
	if err != nil || port <= 0 {
		return fail("PORT: must be a positive integer")
	}

	coupon := pick(raw, "COUPON_MEDIA_URL", "https://www.w3.org/WAI/ER/tests/xhtml/testfiles/resources/pdf/dummy.pdf")
	if _, err := url.ParseRequestURI(coupon); err != nil {
		return fail("COUPON_MEDIA_URL: must be a valid URL")
	}

	menuMode := pick(raw, "WHATSMEOW_MENU_MODE", "text")
	switch menuMode {
	case "text", "buttons", "list":
	default:
		return fail("WHATSMEOW_MENU_MODE: must be text, buttons, or list")
	}

	minMS, err := parseInt(pick(raw, "HUMANIZE_MIN_MS", "20000"))
	if err != nil || minMS < 0 {
		return fail("HUMANIZE_MIN_MS: must be a non-negative integer")
	}
	maxMS, err := parseInt(pick(raw, "HUMANIZE_MAX_MS", "45000"))
	if err != nil || maxMS < 0 {
		return fail("HUMANIZE_MAX_MS: must be a non-negative integer")
	}
	if maxMS < minMS {
		return fail("HUMANIZE_MAX_MS: HUMANIZE_MAX_MS must be >= HUMANIZE_MIN_MS")
	}

	goEnv := pick(raw, "GO_ENV", "")
	if goEnv == "" {
		goEnv = pick(raw, "NODE_ENV", "")
	}

	secret := strings.TrimSpace(pick(raw, "WHATSMEOW_WEBHOOK_SECRET", ""))

	if goEnv == "production" {
		if minMS < 20_000 {
			return fail("HUMANIZE_MIN_MS: HUMANIZE_MIN_MS must be >= 20000 when GO_ENV/NODE_ENV is production")
		}
		if maxMS < 45_000 {
			return fail("HUMANIZE_MAX_MS: HUMANIZE_MAX_MS must be >= 45000 when GO_ENV/NODE_ENV is production")
		}
		if secret == "" {
			return fail("WHATSMEOW_WEBHOOK_SECRET: must be non-empty when GO_ENV/NODE_ENV is production")
		}
	}

	return EnvParseResult{
		OK: true,
		Env: AppEnv{
			StoreDir:       storeDir,
			Port:           port,
			CouponMediaURL: coupon,
			MenuMode:       menuMode,
			WebhookSecret:  secret,
			GoEnv:          goEnv,
			HumanizeMinMS:  minMS,
			HumanizeMaxMS:  maxMS,
		},
	}
}

// ParseProcessEnv - reads os.Getenv keys used by ParseEnv.
func ParseProcessEnv() EnvParseResult {
	keys := []string{
		"WHATSMEOW_STORE_DIR", "PORT", "COUPON_MEDIA_URL", "WHATSMEOW_MENU_MODE",
		"HUMANIZE_MIN_MS", "HUMANIZE_MAX_MS", "WHATSMEOW_WEBHOOK_SECRET", "GO_ENV", "NODE_ENV",
	}
	raw := make(map[string]string, len(keys))
	for _, key := range keys {
		if value, ok := os.LookupEnv(key); ok {
			raw[key] = value
		}
	}
	return ParseEnv(raw)
}

func pick(raw map[string]string, key, fallback string) string {
	if raw == nil {
		return fallback
	}
	if value, ok := raw[key]; ok && value != "" {
		return value
	}
	return fallback
}

func parseInt(value string) (int, error) {
	n, err := strconv.Atoi(strings.TrimSpace(value))
	if err != nil {
		return 0, err
	}
	return n, nil
}

func fail(message string) EnvParseResult {
	return EnvParseResult{OK: false, Error: fmt.Sprintf("Invalid environment configuration: %s", message)}
}

File: WhatsApp-agents/waha/src/contracts/env.ts
//Mariano Montini ('bosque', 'bosquestudio')
import { z } from 'zod'

// Optional webhook secret - empty string treated as unset (open webhook).
const optionalSecretSchema = z.preprocess((value) => {
  if (typeof value !== 'string') return value
  const trimmed = value.trim()
  return trimmed.length === 0 ? undefined : trimmed
}, z.string().min(1).optional())

// Env field schema - Waha credentials, port, coupon URL, menu mode, humanize delays.
export const envSchema = z
  .object({
    WAHA_API_URL: z.string().url('WAHA_API_URL must be a valid URL'),
    WAHA_API_KEY: z.string().min(1, 'WAHA_API_KEY is required'),
    WAHA_SESSION: z.string().min(1, 'WAHA_SESSION is required'),
    PORT: z.coerce.number().int().positive().default(3002),
    COUPON_MEDIA_URL: z
      .string()
      .url()
      .default('https://www.w3.org/WAI/ER/tests/xhtml/testfiles/resources/pdf/dummy.pdf'),
    WAHA_MENU_MODE: z.enum(['text', 'buttons', 'list']).default('text'),
    WAHA_WEBHOOK_SECRET: optionalSecretSchema,
    NODE_ENV: z.string().optional(),
    HUMANIZE_MIN_MS: z.coerce.number().int().nonnegative().default(20_000),
    HUMANIZE_MAX_MS: z.coerce.number().int().nonnegative().default(45_000),
  })
  .superRefine((data, ctx) => {
    if (data.HUMANIZE_MAX_MS < data.HUMANIZE_MIN_MS) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ['HUMANIZE_MAX_MS'],
        message: 'HUMANIZE_MAX_MS must be >= HUMANIZE_MIN_MS',
      })
    }

    // Production delay floor - refuse 0 ms bypass when NODE_ENV is production.
    if (data.NODE_ENV === 'production') {
      if (data.HUMANIZE_MIN_MS < 20_000) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          path: ['HUMANIZE_MIN_MS'],
          message: 'HUMANIZE_MIN_MS must be >= 20000 when NODE_ENV is production',
        })
      }
      if (data.HUMANIZE_MAX_MS < 45_000) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          path: ['HUMANIZE_MAX_MS'],
          message: 'HUMANIZE_MAX_MS must be >= 45000 when NODE_ENV is production',
        })
      }
    }
  })

// Parsed env type - validated configuration used across layers.
export type AppEnv = z.infer<typeof envSchema>

// Env parse result - either validated config or a clear missing-keys error.
export type EnvParseResult =
  | { ok: true; env: AppEnv }
  | { ok: false; error: string }

// Parse process env - fails fast listing missing/invalid keys for startup.
export function parseEnv(raw: NodeJS.ProcessEnv = process.env): EnvParseResult {
  const result = envSchema.safeParse(raw)
  if (!result.success) {
    const details = result.error.issues.map((issue) => `${issue.path.join('.')}: ${issue.message}`).join('; ')
    return { ok: false, error: `Invalid environment configuration: ${details}` }
  }
  return { ok: true, env: result.data }
}
